diff --git /etc/one/oned.conf /etc/one/oned.conf
index be02d646a8..27f876ec36 100644
--- /etc/one/oned.conf
+++ /etc/one/oned.conf
@@ -481,7 +481,7 @@ VM_MAD = [
     NAME           = "kvm",
     SUNSTONE_NAME  = "KVM",
     EXECUTABLE     = "one_vmm_exec",
-    ARGUMENTS      = "-t 15 -r 0 kvm -p",
+    ARGUMENTS      = "-t 15 -r 0 kvm -p -l deploy=deploy.vitastor",
     DEFAULT        = "vmm_exec/vmm_exec_kvm.conf",
     TYPE           = "kvm",
     KEEP_SNAPSHOTS = "yes",
@@ -592,7 +592,7 @@ VM_MAD = [
 
 TM_MAD = [
     EXECUTABLE = "one_tm",
-    ARGUMENTS = "-t 15 -d dummy,lvm,shared,fs_lvm,fs_lvm_ssh,qcow2,ssh,ceph,dev,vcenter,iscsi_libvirt"
+    ARGUMENTS = "-t 15 -d dummy,lvm,shared,fs_lvm,fs_lvm_ssh,qcow2,ssh,ceph,vitastor,dev,vcenter,iscsi_libvirt"
 ]
 
 #*******************************************************************************
@@ -612,7 +612,7 @@ TM_MAD = [
 
 DATASTORE_MAD = [
     EXECUTABLE = "one_datastore",
-    ARGUMENTS  = "-t 15 -d dummy,fs,lvm,ceph,dev,iscsi_libvirt,vcenter,restic,rsync -s shared,ssh,ceph,fs_lvm,fs_lvm_ssh,qcow2,vcenter"
+    ARGUMENTS  = "-t 15 -d dummy,fs,lvm,ceph,vitastor,dev,iscsi_libvirt,vcenter,restic,rsync -s shared,ssh,ceph,vitastor,fs_lvm,fs_lvm_ssh,qcow2,vcenter"
 ]
 
 #*******************************************************************************
@@ -1050,6 +1050,9 @@ INHERIT_DATASTORE_ATTR  = "VCENTER_DS_IMAGE_DIR"
 INHERIT_DATASTORE_ATTR  = "VCENTER_DS_VOLATILE_DIR"
 INHERIT_DATASTORE_ATTR  = "VCENTER_INSTANCE_ID"
 
+INHERIT_DATASTORE_ATTR  = "VITASTOR_CONF"
+INHERIT_DATASTORE_ATTR  = "IMAGE_PREFIX"
+
 INHERIT_IMAGE_ATTR      = "DISK_TYPE"
 INHERIT_IMAGE_ATTR      = "VCENTER_ADAPTER_TYPE"
 INHERIT_IMAGE_ATTR      = "VCENTER_DISK_TYPE"
@@ -1180,6 +1183,14 @@ TM_MAD_CONF = [
     CLONE_TARGET_SHARED = "SELF", DISK_TYPE_SHARED = "RBD"
 ]
 
+TM_MAD_CONF = [
+    NAME = "vitastor", LN_TARGET = "NONE", CLONE_TARGET = "SELF", SHARED = "YES",
+    DS_MIGRATE = "NO", DRIVER = "raw", ALLOW_ORPHANS="format",
+    TM_MAD_SYSTEM = "ssh,shared", LN_TARGET_SSH = "SYSTEM", CLONE_TARGET_SSH = "SYSTEM",
+    DISK_TYPE_SSH = "FILE", LN_TARGET_SHARED = "NONE",
+    CLONE_TARGET_SHARED = "SELF", DISK_TYPE_SHARED = "FILE"
+]
+
 TM_MAD_CONF = [
     NAME = "iscsi_libvirt", LN_TARGET = "NONE", CLONE_TARGET = "SELF", SHARED = "YES",
     DS_MIGRATE = "NO", DRIVER = "raw"
@@ -1219,9 +1230,16 @@ DS_MAD_CONF = [
     NAME = "ceph",
     REQUIRED_ATTRS = "DISK_TYPE,BRIDGE_LIST",
     PERSISTENT_ONLY = "NO",
     MARKETPLACE_ACTIONS = "export"
 ]
 
+DS_MAD_CONF = [
+    NAME = "vitastor",
+    REQUIRED_ATTRS = "DISK_TYPE,BRIDGE_LIST",
+    PERSISTENT_ONLY = "NO",
+    MARKETPLACE_ACTIONS = "export"
+]
+
 DS_MAD_CONF = [
     NAME = "dev", REQUIRED_ATTRS = "DISK_TYPE", PERSISTENT_ONLY = "YES"
 ]
